File: backend/app/schemas/explain.py
"""
Pydantic v2 schemas for AI explanation endpoints.

Depends on: pydantic
Used by: api/v1/risk.py, api/v1/markowitz.py, api/v1/stress.py
"""

from typing import Literal

from pydantic import BaseModel, Field


# ── Shared ──


class ExplanationResponse(BaseModel):
    """AI-generated explanation text."""

    explanation: str = Field(description="Generated explanation in French")


# ── Monte Carlo ──


class ExplainMonteCarloRequest(BaseModel):
    """Request body for Monte Carlo explanation."""

    mode: Literal["beginner", "expert"] = Field(description="UI mode")
    mean_final_value: float
    var_95: float
    probability_of_loss: float
    n_simulations: int = 10_000
    n_days: int = 252


# ── Distribution ──


class ExplainDistributionRequest(BaseModel):
    """Request body for loss distribution explanation."""

    mode: Literal["beginner", "expert"] = Field(description="UI mode")
    var_95: float
    mean_final_value: float
    std_final_value: float
    percentile_5: float
    percentile_95: float


# ── Markowitz ──


class ExplainMarkowitzRequest(BaseModel):
    """Request body for Markowitz frontier position explanation."""

    mode: Literal["beginner", "expert"] = Field(description="UI mode")
    current_sharpe: float
    current_volatility: float
    current_return: float
    max_sharpe_ratio: float
    max_sharpe_volatility: float
    max_sharpe_return: float
    min_variance_volatility: float


# ── Stress ──


class ExplainStressScenario(BaseModel):
    """Single scenario for stress explanation request."""

    scenario_name: str
    total_return: float
    max_drawdown: float
    recovery_days: int | None


class ExplainStressRequest(BaseModel):
    """Request body for stress test explanation."""

    mode: Literal["beginner", "expert"] = Field(description="UI mode")
    scenarios: list[ExplainStressScenario] = Field(min_length=1)


# ── Generic metric ──


class ExplainMetricRequest(BaseModel):
    """Request body for generic metric explanation."""

    metric_name: str = Field(description="Machine-readable metric key (e.g. var_95)")
    metric_value: float = Field(description="Current numeric value of the metric")
    portfolio_id: str = Field(description="Portfolio UUID for context")
    mode: Literal["beginner", "expert"] = Field(description="UI mode")
    context: dict[str, float | str | int | None] = Field(
        default_factory=dict,
        description="Additional context values (e.g. other metrics for comparison)",
    )

File: frontend/src/lib/api/explain.ts
/**
 * TanStack Query hooks for AI chart explanation endpoints.
 *
 * Uses useMutation — explanations are triggered on user click,
 * not automatically on page load.
 *
 * Depends on: @tanstack/react-query, lib/api/client.ts, types/explain.ts
 * Used by: components/shared/ai-chart-explanation.tsx, chart components
 */

import { useMutation } from "@tanstack/react-query";

import { apiClient } from "@/lib/api/client";
import type { ExplanationResponse } from "@/types/explain";
import { useLocaleStore, type Locale } from "@/lib/store/locale-store";

type Mode = "beginner" | "expert";

/** Helper to get the current locale for API calls (non-hook context). */
export function getCurrentLocale(): Locale {
  return useLocaleStore.getState().locale;
}

// ── API functions ──

export async function fetchMonteCarloExplanation(params: {
  mode: Mode;
  mean_final_value: number;
  var_95: number;
  probability_of_loss: number;
  n_simulations: number;
  n_days: number;
  locale?: string;
}): Promise<ExplanationResponse> {
  return apiClient<ExplanationResponse>("/api/v1/risk/explain-montecarlo", {
    method: "POST",
    body: { ...params, locale: params.locale ?? getCurrentLocale() },
  });
}

export async function fetchDistributionExplanation(params: {
  mode: Mode;
  var_95: number;
  mean_final_value: number;
  std_final_value: number;
  percentile_5: number;
  percentile_95: number;
  locale?: string;
}): Promise<ExplanationResponse> {
  return apiClient<ExplanationResponse>("/api/v1/risk/explain-distribution", {
    method: "POST",
    body: { ...params, locale: params.locale ?? getCurrentLocale() },
  });
}

export async function fetchMarkowitzExplanation(params: {
  mode: Mode;
  current_sharpe: number;
  current_volatility: number;
  current_return: number;
  max_sharpe_ratio: number;
  max_sharpe_volatility: number;
  max_sharpe_return: number;
  min_variance_volatility: number;
  locale?: string;
}): Promise<ExplanationResponse> {
  return apiClient<ExplanationResponse>(
    "/api/v1/markowitz/explain-position",
    { method: "POST", body: { ...params, locale: params.locale ?? getCurrentLocale() } },
  );
}

export async function fetchStressExplanation(params: {
  mode: Mode;
  scenarios: {
    scenario_name: string;
    total_return: number;
    max_drawdown: number;
    recovery_days: number | null;
    optimized_drawdown?: number | null;
  }[];
  locale?: string;
}): Promise<ExplanationResponse> {
  return apiClient<ExplanationResponse>("/api/v1/stress/explain-result", {
    method: "POST",
    body: { ...params, locale: params.locale ?? getCurrentLocale() },
  });
}

// ── Generic metric explanation (plain async, not a hook) ──

export async function fetchMetricExplanation(params: {
  metric_name: string;
  metric_value: number;
  portfolio_id: string;
  mode: Mode;
  locale?: string;
  context?: Record<string, number | string | null>;
}): Promise<string> {
  const res = await apiClient<ExplanationResponse>("/api/v1/risk/explain-metric", {
    method: "POST",
    body: { ...params, locale: params.locale ?? getCurrentLocale() },
  });
  return res.explanation;
}

// ── Hooks (useMutation — triggered on click, not on load) ──

export function useMonteCarloExplanation() {
  return useMutation({
    mutationFn: fetchMonteCarloExplanation,
  });
}

export function useDistributionExplanation() {
  return useMutation({
    mutationFn: fetchDistributionExplanation,
  });
}

export function useMarkowitzExplanation() {
  return useMutation({
    mutationFn: fetchMarkowitzExplanation,
  });
}

export function useStressExplanation() {
  return useMutation({
    mutationFn: fetchStressExplanation,
  });
}
